### ISSUE TITLE
Are 3DS2 credit cards are not supported in the V3 version of the api?

### ISSUE THREAD

**Issue by Erik8309** [recurly/recurly-client-net#453]
When looking at the documentation of for the V3 version of the api I cannot see any endpoint that has a body with a property called "three_d_secure_action_result_token_id". 

Nor do we get a token on property "three_d_secure_action_token_id" to supply to the user when we try to use credit cards requiring 3DS2 authentication.

**Comment by drewish** [comment]
We haven't yet finished adding support to V3. I'm not sure of the exact timeline for release but it's in progress now.

**Comment by neuronicnobody** [comment]
@Erik8309 We have an aspirational target to release in Q4 of this year pending other priorities currently in the queue.

Please note the EBA did recently publish an opinion which sets the new PSD2 deadline to 31 December 2020 and prescribes expected actions to be taken during the migration period.

https://eba.europa.eu/eba-publishes-opinion-on-the-deadline-and-process-for-completing-the-migration-to-strong-customer-authentication-sca-for-e-commerce-card-based-payment

**Comment by Erik8309** [comment]
Good to know. I suggest adding it here: https://dev.recurly.com/page/recurly-3d-secure-2-integration-guide
before anyone else spends time on trying to get it to work :)

Im completely new to having payment in my projects, but from a user point of view: 3ds2 (or similar) has been active in Sweden for years. You get a popup saying that you have to provide additional authentication. This is provied using different forms of 2-factor.

**Comment by neuronicnobody** [comment]
@Erik8309 We've been keeping any information we know regarding the directive itself and any relevant dates here:
https://docs.recurly.com/docs/revised-payment-services-directive-psd2

The integration guide is primarily focused on step by step instructions for completing the actual 3DS2 implementation.

For the previous 3DS experiences you've seen, it's probably 3DS version 1. The 3DS2 spec was released along with the PSD2. You can read more about it here: https://docs.recurly.com/docs/revised-payment-services-directive-psd2

That being said, releasing 3DS2 support in the Recurly v3 API is on our roadmap and close to completion. Final code review and QA are the last steps to complete.

**Comment by peturh** [comment]
There is still issues with starting the "normal" 3DS flow. Since that expects the same sort of token (`three_d_secure_action_result_token_id`), which is not in the API as of now?

**Comment by oskma** [comment]
Should this work now in the 3.1 release of the api client? I can't get the three_d_secure_action_token_id but I'm not sure if I am doing something wrong or if it is not supported yet.

**Comment by drewish** [comment]
@oskma The support should be there now. Here's the property on the TransactionError:
https://github.com/recurly/recurly-client-net/blob/c74b5c04ba2a73c356ad9fd587809221451ca8b3/Recurly/Resources/TransactionError.cs#L34-L36

It'll only have a value if a 3DS action is required though. 

Could you say more about how you're trying to test this? Maybe show some of your code?

**Comment by oskma** [comment]
I get a token for 4000-0000-0000-3220 and then I send in that token to Recurly.Client.CreateSubscription. Then I get an exception telling me "Your card must be authenticated with 3-D Secure before continuing."

Is that the correct flow? Because I don’t understand how to get the TransactionError recourse from that exception. 

I can build a simple example that calls CreateSubscription if that would help?

**Comment by drewish** [comment]
Yeah if you could show how you're passing it into create subscription that would be helpful. Also if you can include the request id from the error on one of these requests I can check our logs and see what's actually making it over to our side.

**Comment by oskma** [comment]
Here is test I built

```c#
[Test]
public void CreateSubscription()
{
    var subscriptionCreate = new SubscriptionCreate
    {
        Account = new AccountCreate() { Email = "test@test.com", Code = Guid.NewGuid().ToString() },
        PlanCode = "notrialplan",
        UnitAmount = 1,
        Currency = "EUR",
    };

    var response = new HttpClient().GetAsync(
        $"https://api.recurly.com/js/v1/token?number=4000-0000-0000-3220&month=12&year=21&first_name=Firstname&last_name=Lastname&address1=Street&city=City&postal_code=12345&country=Sweden&key={_recurlyConfig.PublicKey}")
        .Result;

    dynamic responseObject = response.Content.ReadAsAsync<System.Dynamic.ExpandoObject>().Result;

    subscriptionCreate.Account.BillingInfo = new BillingInfoCreate
    {
        TokenId = responseObject.id
    };

    try
    {
        _recurlyClient.CreateSubscription(subscriptionCreate);
    }
    catch (RecurlyError ex)
    {
        Console.Write(Newtonsoft.Json.JsonConvert.SerializeObject(ex));
    }
}
```
and here is the exception:

```json
{ 
   "Error":{ 
      "message":"Your card must be authenticated with 3-D Secure before continuing.",
      "params":null,
      "type":"transaction"
   },
   "Message":"Your card must be authenticated with 3-D Secure before continuing.",
   "Data":{ 

   },
   "InnerException":null,
   "StackTrace":"   at Recurly.BaseClient.HandleResponse(IRestResponse resp)\r\n   at Recurly.BaseClient.MakeRequest[T](Method method, String url, Request body, Dictionary`2 queryParams)\r\n   at Recurly.Client.CreateSubscription(SubscriptionCreate body)\r\n   
at Implementation.Tests.SubscriptionServiceTests.CreateSubscription_3DS2() in C:\\Work\\tests\\Implementation.Tests\\SubscriptionServiceTests.cs:line 518",
   "HelpLink":null,
   "Source":"Recurly",
   "HResult":-2146233088
}
```

**Comment by Herlix** [comment]
Hi! 
Through debugging I found that the code is missing because it's not being picked up during de-serialization. The error object used is missing the field needed. However there is a object [https://github.com/recurly/recurly-client-net/blob/master/Recurly/Resources/ErrorMayHaveTransaction.cs](ErrorMayHaveTransaction) but it is not used.

**Comment by drewish** [comment]
Ah okay I understand now. We're not instantiating the right class so it doesn't matter if we've added the attribute. I'd worked on the server side part of this so I'll circle up with the folks that know .net and see how they'd like to address this.

**Comment by bhelx** [comment]
@Herlix I'm working on a fix for that this week. I'll ping back here when it's ready.

**Comment by bhelx** [comment]
@Herlix If you have a moment, would you confirm this PR fixes the problem for you?

https://github.com/recurly/recurly-client-net/pull/465

Assuming it does, we will test it and get a release out for you immediately.

**Comment by Herlix** [comment]
> @Herlix If you have a moment, would you confirm this PR fixes the problem for you?
> 
> #465
> 
> Assuming it does, we will test it and get a release out for you immediately.

It looks good :)

**Comment by bhelx** [comment]
@Herlix (et al): this was just released as 3.1.1 https://github.com/recurly/recurly-client-net/releases/tag/3.1.1

Apologies for the problem. Please let us know if you find anything else.

**Comment by Herlix** [comment]
Sounds great! Thank you :)
